logdna/resource_view_test.go

```diff
@@ -181,6 +181,35 @@ func TestViewBasic(t *testing.T) {
 	})
 }
 
+func TestViewBasicUpdate(t *testing.T) {
```

Comment:
So in these tests, I don't see any expected failure tests&mdash;explicit failure cases such as passing a boolean instead of a string or an invalid value instead of a valid one for values that are constrained to a set. I'm not comfortable approving a PR without those, personally, as those failure cases need to be handled gracefully to ensure a strong developer experience.

Reply:
@nimbinatus all the expected failure tests can be found here: https://github.com/logdna/terraform-provider-logdna/blob/8f89c85fe571dd54e432558ad8edbe4440e9a031/logdna/resource_view_test.go#L12

Reply:
I see. I don't see a failure test for a view, however. Maybe this isn't a thing in Go, but I suspect it is. I'm expecting something like this, but in actual Go code and not my poor pseudocode:
```go
func TestAssertErrorViewBulkEmails(t *testing.T err error) {
	name := "test"
	query := "test"
	category := "DEMO"
	tags1 := "host1"
	tags2 := 123

	resource.Test(t, resource.TestCase{
		Providers: testAccProviders,
		Steps: []resource.TestStep{
			{
				Config: testViewConfigBulkEmails(name, query, category, tags1, tags2),
				Check: resource.ComposeTestCheckFunc(
					testViewExists("logdna_view.new"),
					resource.TestCheckResourceAttr("logdna_view.new", "name", name),
					resource.TestCheckResourceAttr("logdna_view.new", "query", query),
					resource.TestCheckResourceAttr("logdna_view.new", "category.#", "1"),
					resource.TestCheckResourceAttr("logdna_view.new", "category.0", category),
					resource.TestCheckResourceAttr("logdna_view.new", "tags.#", "2"),
					resource.TestCheckResourceAttr("logdna_view.new", "tags.0", tags1),
					resource.TestCheckResourceAttr("logdna_view.new", "tags.1", tags2),
					resource.TestCheckResourceAttr("logdna_view.new", "webhook_channels.#", "0"),
				),
			},
		},
	})
        if assert.Error(t, err) {
	      assert.Equal(t, expectedError, err)
        }
        else {
            t.Fail()
	    t.Logf("View created with invalid values %s", tags2) # Though I'd also expect the single string for `categories` to fail as it's not an array.
        }
}
```

Personally, I'd also add tests to validate each string for the key-value pairs that have set values as options. For example, what if someone put `pagerduty_channel.0.immediate` as `"no"` instead of `"false"`&mdash;or if they mistyped and entered `"fasle"`, which I've done myself many times? How do you test for a graceful failure or error here? We'd rather see that kind of mistake just error out with a reasonable message for the user versus possibly crashing or giving the user no feedback at all.

Reply:
@nimbinatus am following the guidelines provided here on how to test error handling with the terraform sdk: https://www.terraform.io/docs/extend/best-practices/testing.html#expecting-errors-or-non-empty-plans:

You can see in the test here is the user doesn't provide a servicekey the provider will give the following error:

`
ExpectError: regexp.MustCompile("The argument \"servicekey\" is required, but no definition was found."),
`

Reply:
also for the situation as described you'll get the following error: 

![Screen Shot 2020-10-26 at 6 44 22 PM](https://user-images.githubusercontent.com/10788080/97246826-86a1b580-17bb-11eb-8f1c-a9ec46453656.png)

Am not able to assert on the expected on an error of that sort with the terraform sdk.

Reply:
@dm36 RE that screenshot, it looks like you'd just need to adjust the regex to match on the error you're looking for? It looks like it should work to use `ExpectError` in that way and testing that the user gets the expected error output - the extra debug log output in the error text might be throwing you off. Can that debug output be disabled under test?

you also likely need to escape (at minimum) the bracket characters in the `regexp.MustCompile`
`Error: "channels[0].immediate" must be...`

Reply:
Yeah I'm not sure how to make assertions on errors returned from the config-api with the terraform sdk and the ExpectError function. I've also created a post here: https://discuss.hashicorp.com/t/testing-errors-when-building-a-terraform-provider/16770

Reply:
can configure terraform error levels but even the lowest level still has that debug message at the beginning: https://www.terraform.io/docs/internals/debugging.html

![Screen Shot 2020-10-27 at 11 34 31 PM](https://user-images.githubusercontent.com/10788080/97400824-fe9bd880-18ac-11eb-9b02-d9a22b62bc9c.png)

Reply:
After talking with some folks at Hashicorp, the `ExpectError` is the right thing to use. For the most part, people test for the happy case, but it isn't necessarily uncommon to test for errors as I noted. They sent along samples like https://github.com/terraform-providers/terraform-provider-aws/search?q=expecterror and https://github.com/hashicorp/terraform-provider-google/search?q=expecterror.

I think that, in this case, the `ExpectError` is the assert, so it is the way to go. Some similar ideas to look at:
- https://github.com/hashicorp/terraform-provider-google/blob/b0e3ae2a9e7235af180fc1c06380659d0b3a474f/google/resource_bigtable_instance_test.go#L27
- A bit more involved, using the `TestCheckFunc` ([docs](https://godoc.org/github.com/hashicorp/terraform-plugin-sdk/helper/resource#TestCheckFunc)) [here](https://github.com/terraform-providers/terraform-provider-aws/blob/a62eec8d6752388cebbb62595bc037318de23286/aws/resource_aws_ses_domain_identity_verification_test.go#L81-L128) called [here](https://github.com/terraform-providers/terraform-provider-aws/blob/a62eec8d6752388cebbb62595bc037318de23286/aws/resource_aws_ses_domain_identity_verification_test.go#L39)

Reply:
I'm also okay with dropping this for a later PR if @mdeltito is good with that.

Reply:
yeah I think this is good - we have test coverage for error cases on [custom validators](https://github.com/logdna/terraform-provider-logdna/blob/9035ef495c2405f07b99115a3905eb6dba5571b5/logdna/resource_view_test.go#L94-L116) (among other things) which I think is the most important. 

@dm36 we'll need to sync up on what the issue was with using `ExpectError` for other error cases at a later point, I'm not sure I really understand based on your screenshots after testing these things locally.